cw/capture.py

```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
+    if capture_cfg["curve"] == 'p256':
+        curve_order_n = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
+        # key_bytes = 256 // 8
+        seed_bytes = 320 // 8
+    else:
+        # TODO: add support for P384
+        raise NotImplementedError(f'Curve {capture_cfg["curve"]} is not supported')
+
+    # Check the lengths in the key/plaintext generator. In this case, "key"
+    # means seed and "plaintext" means mask.
+    if ktp.keyLen() != seed_bytes:
+        raise ValueError(f'Unexpected seed length: {ktp.keyLen()}.\n'
+                         f'Hint: set key len={seed_bytes} in the configuration file.')
+    if ktp.textLen() != seed_bytes:
+        raise ValueError(f'Unexpected mask length: {ktp.textLen()}.\n'
+                         f'Hint: set plaintext len={seed_bytes} in the configuration file.')
+
+    # Seed the RNG and generate a random fixed seed for all traces of the
+    # keygen operation.
+    seed_fixed = ktp.next_key()
+    print(f'fixed seed = {seed_fixed.hex()}')
+    if len(seed_fixed) != seed_bytes:
+        raise ValueError(f'Fixed seed length is {len(seed_fixed)}, expected {seed_bytes}')
+
+    # Expected key is `seed mod n`, where n is the order of the curve and
+    # `seed` is interpreted as little-endian.
+    expected_fixed_key = int.from_bytes(seed_fixed, byteorder='little') % curve_order_n
+
+    # register ctrl-c handler to not loose already recorded traces if measurement is aborted
+    signal.signal(signal.SIGINT, partial(abort_handler, project))
+
+    sample_fixed = 1
+    # Loop to collect each power trace
+    for _ in tqdm(range(capture_cfg["num_traces"]), desc='Capturing',
+                  ncols=80):
+
+        if reset_firmware:
+            ot.program_target(fw_bin, pll_frequency)
+
+        ot.scope.adc.offset = capture_cfg["offset"]
+
+        if capture_cfg["masks_off"] is True:
+            # Use a constant mask for each trace
+            mask = bytearray(capture_cfg["plain_text_len_bytes"])  # all zeros
+        else:
+            # Generate a new random mask for each trace.
+            mask = ktp.next_text()
+        tqdm.write("Starting new trace....")
+        tqdm.write(f'mask   = {mask.hex()}')
+
+        if sample_fixed:
+            # Set the seed.
+            ot.target.simpleserial_write('x', seed_fixed)
+            tqdm.write(f'seed   = {seed_fixed.hex()}')
+
+            # Check for errors.
+            err = ot.target.read()
+            if err:
+                raise RuntimeError(f'Error writing seed: {err}')
+
+            # Arm the scope
+            ot.scope.arm()
+
+            # Send the mask and start the keygen operation.
+            ot.target.simpleserial_write('k', mask)
+
+            # Wait until operation is done.
+            ret = ot.scope.capture(poll_done=True)
+            if ret:
+                raise RuntimeError('Timeout during capture')
+
+            # Check the number of cycles where the trigger signal was high.
```

Comment:
Should we compare it with an expected number instead of printing?

Reply:
We could do that, however the actual number of cycles varies by `+-2`  cycles.
If needed, let's do that in another PR.